File: crates/keystone/tpt_sdk/lib/src/storage.dart
/// Offline cache storage. Mirrors the SharedPreferences shape (string values
/// keyed by string) so callers can plug in `shared_preferences` directly. The
/// default [InMemoryStorage] is not persisted across app restarts.
abstract class Storage {
  Future<String?> getString(String key);
  Future<void> setString(String key, String value);
  Future<void> remove(String key);
}

/// Non-persistent in-memory [Storage] used when no adapter is supplied.
class InMemoryStorage implements Storage {
  final Map<String, String> _store = {};

  @override
  Future<String?> getString(String key) async => _store[key];

  @override
  Future<void> setString(String key, String value) async => _store[key] = value;

  @override
  Future<void> remove(String key) async => _store.remove(key);
}

File: crates/keystone/tpt-sdk-android/src/main/kotlin/tpt/sdk/android/Storage.kt
package tpt.sdk.android

import android.content.SharedPreferences
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Offline-cache storage. Default [SharedPreferencesStorage] persists across
 *  app restarts; [InMemoryStorage] does not. */
interface Storage {
    suspend fun getString(key: String): String?
    suspend fun putString(key: String, value: String)
    suspend fun remove(key: String)
}

/** SharedPreferences-backed [Storage] (Android's standard key-value store). */
class SharedPreferencesStorage(private val prefs: SharedPreferences) : Storage {
    override suspend fun getString(key: String): String? = prefs.getString(key, null)
    override suspend fun putString(key: String, value: String) {
        prefs.edit().putString(key, value).apply()
    }
    override suspend fun remove(key: String) {
        prefs.edit().remove(key).apply()
    }
}

/** Non-persistent in-memory [Storage], handy for previews and unit tests. */
class InMemoryStorage : Storage {
    private val store = mutableMapOf<String, String>()
    override suspend fun getString(key: String): String? = store[key]
    override suspend fun putString(key: String, value: String) {
        store[key] = value
    }
    override suspend fun remove(key: String) {
        store.remove(key)
    }
}

@Serializable
private data class CacheWrapper(val result: QueryResult, val ts: Long)

internal fun cacheKey(query: String, params: List<Any?>): String =
    "tpt:" + Json.encodeToString(
        mapOf("q" to query, "p" to params.map { it.toString() }),
    )

internal fun decodeCacheEntry(raw: String?, ttlMs: Long): QueryResult? {
    if (raw == null) return null
    return try {
        val wrapper = Json.decodeFromString<CacheWrapper>(raw)
        if (System.currentTimeMillis() - wrapper.ts > ttlMs) null else wrapper.result
    } catch (_: Exception) {
        null
    }
}

internal fun encodeCacheEntry(result: QueryResult, ttlClockMs: Long): String =
    Json.encodeToString(CacheWrapper(result, ttlClockMs))
